### ISSUE TITLE
Groovy scripting: java.lang.ClassFormatError: Illegal class name

### ISSUE THREAD

**Issue by richardhundt** [elastic/elasticsearch#14628]
Hi there,

the Groovy scripting service generates a hex sha1 digest of the script content which is used as a class name during compilation.

In version 1.7 (and probably current as well), if the digest happens to start with a digit, then scripts fail to run, throwing a ClassFormatError. I've only observed this with script files (i.e. under config/scripts).

My current workaround involves adding arbitrary comments and white space to the script until its content hash starts with a word character. An easy fix would be simply to prepend a valid character (`_` or whatever) to the `fake` class name.

Cheers,
Richard

**Comment by s1monw** [comment]
oh boy, do you wanna open a PR to fix this? I also wonder if you have a reproduction for this issue so we can add a test?

**Comment by richardhundt** [comment]
Sure, I'll open a PR and have a test case.

**Comment by s1monw** [comment]
thank you!

**Comment by richardhundt** [comment]
Umm... has groovy scripting been removed from 2.x? I can see it in master, but now I'm not sure whether it'll be around in future. Which branch d'you want me to work against?

**Comment by richardhundt** [comment]
nvm, it seems to have moved to core

**Comment by s1monw** [comment]
> nvm, it seems to have moved to core
yeah it's in `plugins/lang-groovy` now. Please make the PR against master

**Comment by richardhundt** [comment]
I've managed to create a failing test in 1.7, and reproduced in 2.0 but not in master.

Turns out master doesn't use `GroovyClassLoader.parseClass()` anymore. There's a comment [here](https://github.com/elastic/elasticsearch/blob/master/plugins/lang-groovy/src/main/java/org/elasticsearch/script/groovy/GroovyScriptEngineService.java#L174).

So it looks like it's fixed.

For posterity, this only fails if the script defines a method, which is probably why it hasn't bitten many people before.

For example, the following groovy snippet:

```
def answer(a) { return a + 2 }; return answer(40)
```
Hashes to `6e41335324a948a5e324e03445ea9c9f9c19e303` and then I see:

```
..., "reason":"failed to run indexed script [test] using lang [groovy]","caused_by":{"type":"class_format_error","reason":"Illegal class name \"6e41335324a948a5e324e03445ea9c9f9c19e303$answer\" in class file 6e41335324a948a5e324e03445ea9c9f9c19e303$answer"}
```

I have a test in the `messy` namespace, which you might want for regression, but not sure if it's worth polluting your code with.
